daemon: Add LB-only mode (--datapath-mode=lb-only)

brb opened the pull request:
The LB-only mode is intended to be run on load-balancing nodes which have connectivity neither to kube-apiserver nor kvstore. In this mode services are supposed to be configured via Cilium REST API.

Chris has successfully figured out that disabling the identity allocator is enough for cilium-agent to run such on nodes (no connectivity to kube-apiserver/kvstore). Therefore, the LB-only mode disabled it.

In addition, when running in the LB-only mode, cilium-agent will configure some flags to achieve the following:

* Enable kube-proxy replacement's NodePort and ExternalIPs implementation for access from outside, and HostReachableServices - from inside LB node.
* Do not install iptables-rules and disable L7 proxy.
* Disable the tunnel mode.
* Disable health-checking (it's going to be implemented separately).

An example how to run cilium-agent in the LB-only mode:

    # ./daemon/cilium-agent                               \
        --enable-ipv4=true --enable-ipv6=false          \
        --datapath-mode=lb-only --node-port-algorithm=maglev \
        --devices=eth0

Afterwards, a service can be added with:

    # ./cilium/cilium service update                    \
        --frontend '7.7.7.7:80'                         \
        --backends '192.168.34.11:80,192.168.34.12:80'  \
        --id 101 --k8s-external

Keep it mind that the LB-only mode hasn't been optimized yet - e.g. some controllers are unnecessary, and thus can be disabled. This will be addressed in the future. Next step is to introduce e2e tests.

```release-note
Add --datapath-mode=lb which allows cilium-agent to run as a standalone loadbalancer
```

Signed-off-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Add LB-only mode (--datapath-mode=lb-only)

The LB-only mode is intended to be run on load-balancing nodes which
have connectivity neither to kube-apiserver nor kvstore. In this mode
services are supposed to be configured via Cilium REST API.

Chris has successfully figured out that disabling the identity allocator
is enough for cilium-agent to run such on nodes (no connectivity to
kube-apiserver/kvstore). Therefore, the LB-only mode disabled it.

In addition, when running in the LB-only mode, cilium-agent will
configure some flags to achieve the following:

    * Enable kube-proxy replacement's NodePort and ExternalIPs
      implementation for access from outside, and HostReachableServices
      - from inside LB node.
    * Do not install iptables-rules and disable L7 proxy.
    * Disable the tunnel mode.
    * Disable health-checking (it's going to be implemented separately).

An example how to run cilium-agent in the LB-only mode:

    # ./daemon/cilium-agent                                     \
        --enable-ipv4=true --enable-ipv6=false                  \
        --datapath-mode=lb-only --node-port-algorithm=maglev    \
        --devices=eth0

Afterwards, a service can be added with:

    # ./cilium/cilium service update                    \
        --frontend '7.7.7.7:80'                         \
        --backends '192.168.34.11:80,192.168.34.12:80'  \
        --id 101 --k8s-external

Signed-off-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

kkourt commented on pkg/datapath/option/option.go:
```diff
@@ -22,4 +22,7 @@ const (
 	DatapathModeIpvlan = "ipvlan"
+
+	// DatapathModeLB specifies lb-only datapath mode.
+	DatapathModeLB = "lb"
```
Maybe make this: `DatapathModeLbOnly = "lb-only"`? Makes code more greppable/readable I think.

kkourt reviewed (commented):
Cool!
I Appreciate the detailed commit message, thanks :)
Other than a minor comment, LGTM.

fristonio reviewed (approved):
LGTM apart from the change requested by Kornilios.
One question, I believe we can add tests to `Runtime` tests for this feature right? Not a blocker can be taken as a follow-up.

christarazi committed:
k8s: Disable K8s with LB mode

This allows users to run with LB mode with Cilium deployed as a
daemonset without needing to modify our charts. Our charts will inject
the "K8S_NODE_NAME" environment variable which trigger the previous
logic to enable K8s.

Signed-off-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi reviewed (approved):
:tada:

ti-mo reviewed (approved):
Cool, LGTM! I'm missing context about the need for this feature, is there an issue somewhere that describes the practical use case? Is the user expected to implement their own control plane and use the Cilium agent in LB-only mode as the data path?
